luda-editor/client/src/components/sequence_player.rs

```diff
@@ -0,0 +1,312 @@
+    ShowingCut {
+        cut_index: usize,
+    },
+    Transitioning {
+        from_cut_index: usize,
+        transition_progress: OneZero,
+        start_time: Time,
+    },
+}
+
+impl SequencePlayer {
+    pub fn new(sequence: Sequence, project_shared_data: ProjectSharedData) -> Self {
+        Self {
+            sequence,
+            project_shared_data,
+            state: State::ShowingCut { cut_index: 0 },
+        }
+    }
+    pub fn render(&self, props: Props) -> RenderingTree {
+        if self.sequence.cuts.is_empty() {
+            return RenderingTree::Empty;
+        }
+
+        let inner_content_rect = get_inner_content_rect(props.wh);
+
+        translate(
+            inner_content_rect.x(),
+            inner_content_rect.y(),
+            match &self.state {
+                &State::ShowingCut { cut_index } => {
+                    let cut = self.sequence.cuts.get(cut_index).unwrap();
+                    render([
+                        self.render_images(inner_content_rect.wh(), cut, 1.0.one_zero()),
+                        self.render_text_box(inner_content_rect.wh()),
+                        self.render_text(inner_content_rect.wh(), cut),
+                        simple_rect(
+                            inner_content_rect.wh(),
+                            Color::TRANSPARENT,
+                            0.px(),
+                            Color::TRANSPARENT,
+                        )
+                        .attach_event(|builder| {
+                            builder.on_mouse_down_in(|_| {
+                                namui::event::send(InternalEvent::NextCut);
+                            });
+                        }),
+                    ])
+                }
+                &State::Transitioning {
+                    from_cut_index,
+                    transition_progress,
+                    start_time: _start_time,
+                } => render([
+                    self.render_transition(
+                        inner_content_rect.wh(),
+                        from_cut_index,
+                        transition_progress,
+                    ),
+                    self.render_text_box(inner_content_rect.wh()),
+                ]),
+            },
+        )
+    }
+    pub fn update(&mut self, event: &dyn std::any::Any) {
+        match &mut self.state {
+            State::ShowingCut { .. } => {}
+            State::Transitioning {
+                from_cut_index,
+                transition_progress,
+                start_time,
+            } => {
+                let transition_duration = 500.ms();
+
+                let delta_time = Time::now() - *start_time;
+                if delta_time > transition_duration {
+                    self.state = State::ShowingCut {
+                        cut_index: *from_cut_index + 1,
+                    };
+                } else {
+                    *transition_progress = (delta_time / transition_duration).one_zero();
+                }
```

Comment:
It looks good to me, but is there a reason for calculating `transition_progress` in update not render?
`transition_progress` is not updated by specific event, but rather indicates how it will be rendered over time.

Reply:
To make sure transition_progess is always less than 1 on render. If other update job takes too much time, it would be bigger than 1 at render time.
Yeh, it's very rare case but I just thought about it.